### PR TITLE
[Private media] Use exponential backoff when API returns error 429

### PR DESCRIPTION
**Pull request by adamziel**
#### Changes proposed in this Pull Request

This is related to Coming Soon/Private atomic sites project (paObgF-Ui-p2). The rationale for this change is described in D41360-code.

This PR is a Calypso counterpart of D41360-code that makes sure we use a customized retry policy for private media files, if the request returned HTTP status 429.

#### Testing instructions

1. Create a new private atomic site (p58i-8OX-p2)
1. Go to /media
1. Confirm the files are being correctly loaded
1. If you see some gray squares instead of actual thumbnails, confirm they're loaded properly after refreshing the page (that's an API issue fixed in another diff)

### CODE REVIEW

**Comment by adamziel** [comment]
I fully agree that using data-layer here would be the ultimate solution. We even started that way but then hit some roadblocks and found that getting it to work properly in this case would be a medium-sized project on it's own. E.g. I don't think it's possible to differentiate retry strategy depending on the error, or use backoff times like `retry^2/3`. On occasions it would retry infinite amount of times, and debugging didn't yield a clear reason - I've even seen it happening in production on unrelated endpoints a few times. Also I couldn't find any place in our codebase where we're doing exponential backoff ATM. Overall improving data-layer to play nicely with this use-case would be really beneficial, but unfortunately not feasible for the very first version of the project.

**Comment by dmsnell** [comment]
> I don't think it's possible to differentiate retry strategy depending on the error

the retry strategies are fully customizable by supplying a `retryPolicy` to the network request. we've been reluctant to change the behavior based on the error message since it hasn't been clear that there's a need to do that. in every case I've seen so far (except in some cases that were very improperly using a `POST` request) we realized that the default policy was enough.

what behaviors do you want to change?

> or use backoff times like retry^2/3
> I couldn't find any place in our codebase where we're doing exponential backoff ATM

every request running through the data layer will use exponential backoff unless they have manually overridden the default policy. there are also good reasons to avoid having a bunch of different failure-response patterns, or different backoff strategies.

Calypso borrows a technique from Amazon called "[decorrelated jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)" to help make sure that if for some reason a server resource fails that thousands of clients are requesting, then we won't have a storm of requests on the server at the expected backoff delays - it smooths the curve on the server to prevent cascading outages.

> On occasions it would retry infinite amount of times, and debugging didn't yield a clear reason - I've even seen it happening in production on unrelated endpoints a few times.

That's strange indeed. The default policy specifies giving up after three retries, which on average should be around 10s after the first attempt.

Are those requests being polled by some mechanism? Or not going through the data layer?

> Overall improving data-layer to play nicely with this use-case would be really beneficial, but unfortunately not feasible for the very first version of the project.

It's probably a lot easier than you think, but you can move on regardless. Please do try to find another way than `undocumented.js` though since that has a long list of outstanding problems.

If you share what isn't working for you with the data layer after reading this comment I'm more than happy to share what I can think of to resolve the problems. Specifically…

> We even started that way but then hit some roadblocks and found that getting it to work properly in this case would be a medium-sized project on it's own.

Can you share what those roadblocks were?

**Comment by adamziel** [comment]
> the retry strategies are fully customizable by supplying a retryPolicy to the network request

I meant that retryPolicy is plain object and here we'd need to use a callback to differentiate between different errors. But hmm, when I think about it some more we could just use exponential backoff for all errors, there should be no need to differenciate - in which case an object would be enough.

> Calypso borrows a technique from Amazon called "decorrelated jitter" to help make sure that if for some reason a server resource fails that thousands of clients are requesting, then we won't have a storm of requests on the server at the expected backoff delays - it smooths the curve on the server to prevent cascading outages.

Hey this is pretty cool, TIL - thank you for sharing!

> Are those requests being polled by some mechanism? Or not going through the data layer?

I'll be sure to document that the next time it happens.

> every request running through the data layer will use exponential backoff unless they have manually overridden the default policy.

Ah that explains why I didn't find anything by using grep - alright, that's optimistic. 

> Can you share what those roadblocks were?

Mostly infinite retries that were hard to debug, with a sprinkle of other things I mentioned in my previous comment. We've been playing with it in this PR, but couldn't get it just right:

https://github.com/Automattic/wp-calypso/pull/39616/commits/b854bfd3a8c0c1f42551cc9fc5aaa9350ff46bee
 
I am not certain, but I suspect our problems stemmed from the fact that the returned data isn't really in an "envelope" format Calypso requires, but is a blob instead - so even when it's an error you need to use a `FileReader` to grab it's contents instead of just reading JSON.

**Comment by adamziel** [comment]
@dmsnell I just reproduced one of the problems we've encountered the last time:

https://video.wordpress.com/embed/r0vtdehS?hd=0&autoPlay=0&permalink=0&loop=0

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 1]
```diff
@@ -2489,45 +2489,61 @@ Undocumented.prototype.startMigration = function( sourceSiteId, targetSiteId ) {
...
-	const fetchMedia = () => this.wpcom.req.get( { ...params, responseType: 'blob' }, fn );
-
-	if ( ! maxSize ) {
-		return fetchMedia();
-	}
+	return new Promise( ( resolve, _reject ) => {
+		const fetchMedia = () =>
+			this.wpcom.req.get( { ...params, responseType: 'blob' }, ( error, data ) => {
+				if ( error || ! ( data instanceof Blob ) ) {
+					_reject( error );
```
Comment (jblz) [inline client/lib/wpcom-undocumented/lib/undocumented.js:2504]: The non-error / invalid response type case won't be passed through the error chain like this.  Should we pass the invalid response back in these instances? Something like:

```suggestion
				const success = ! error && data instanceof Blob;
				if ( ! success ) {
					_reject( error || { error: 'invalidResponse', data } );
```

**Comment by jblz** [comment]
I'm not sure why that canary is failing, I tried restarting it and it erred out at the same spot.  Maybe a rebase would fix?

**Comment by dmsnell** [comment]
@adamziel @jblz in reviewing the video and the linked commits I think a screenshare would be most helpful on the data layer side: in short, I believe that there are some explanations in the code that uses the data layer causing the behaviors you're seeing.

I'm pretty sure it doesn't support `Blob` responses but I'm also pretty sure that would be a small fix. if you do merge this PR, please swing around and let's have that pairing session to figure this out.

**Comment by dmsnell** [comment]
I left some notes on your linked commit @jblz 

https://github.com/Automattic/wp-calypso/commit/b854bfd3a8c0c1f42551cc9fc5aaa9350ff46bee#r38349034

**Comment by dmsnell** [comment]
@adamziel there are two infinite loops in the code, one is related to the data layer. If I were go guess then I'd suggest that it's purely the absence of `Blob` response handling which causes the loop you see in the network panel.

By not supporting the `Blob` the request fails in `fromApi` which puts it in `error` state.
The app re-renders and triggers a new fetch which repeats this cycle.

Why didn't it fail and stay put? `useEffect()`

`useEffect()` gets re-run whenever `imageObjectUrl` changes, but the `useEffect()` callback _changes `imageObjectUrl`_ creating the loop. From a text perspective it may seem puzzling since you'd think the return URL is the same, but it's not a `string` - it's a `DOMString`, and `URL.createObjectURL()` does just that - it creates an object, which when compared to any other object, even with the same properties, will compare as unequal and trigger `useEffect()` to run again.

But it doesn't stop there either because `createObjectURL()` is creating a reference to memory and each invocation, even with the same input `Blob` will allocate a new reference and therefore the output will be different. This is also known as a memory leak if we keep calling it.

```ts
const b = new Blob(['hi'],{type:'text/plain'});
const u1 = URL.createObjectURL(b); // "blob:https://github.com/046667e7-7c1a-4466-9861-9da14b8a98cb"
const u2 = URL.createObjectURL(b); // "blob:https://github.com/4f7f4405-ccaf-4725-bd45-c1acfabecb94"

u1 !== u2;
u1.toString() !== u2.toString();
```

In other words, the infinite loop is operating as expected since we instructed it to do so 🙃 

In general, please use `useEffect()` with caution and definitely second-guess yourself if you ever think you need it. It can have surprising and unexpected issues.

**Comment by adamziel** [comment]
@dmsnell sounds good! Just a note that that effect used to depend on `imageObjectUrl ` in the past, but it was updated and now it's just `}, [ filePath, requestId, siteSlug ] );` (also on video). These don't change with every request. I traced the code execution a bit and the infinite loop is data layer constantly dispatching the same redux action over and over again. Let's see if we can figure it out together some time soon!

**Comment by adamziel** [comment]
Woo fails in other branches too (https://github.com/Automattic/wp-calypso/pull/40880 https://github.com/Automattic/wp-calypso/pull/40879 https://github.com/Automattic/wp-calypso/pull/40877). I'm going to proceed as these failures are unrelated to this PR (`session not created: This version of ChromeDriver only supports Chrome version 79`).

**Comment by dmsnell** [comment]
> I traced the code execution a bit and the infinite loop is data layer constantly dispatching the same redux action over and over again.

Sounds good. It's very likely failing to parse because of the `Blob` vs. `json`, so the request comes back as a 200, but we fail to parse, so it tries again. That could be wrong, but there's nothing inherent in the data layer that should be causing these loops.
